src/Simplify.cpp

```diff
@@ -729,7 +741,19 @@ class Simplify : public IRMutator {
         } else if (broadcast_a && broadcast_b) {
             // Broadcast + Broadcast
             expr = Broadcast::make(mutate(broadcast_a->value - broadcast_b->value),
-                                 broadcast_a->lanes);
+                                   broadcast_a->lanes);
+        } else if (cast_a && cast_b &&
+                   cast_a->value.type() == cast_b->value.type() &&
+                   cast_a->type.is_int() &&
+                   cast_b->type.is_int() &&
+                   cast_a->value.type().is_int() &&
+                   cast_b->value.type().is_int() &&
+                   no_overflow(cast_a->type) &&
+                   no_overflow(cast_b->type) &&
+                   no_overflow(cast_a->value.type()) &&
+                   no_overflow(cast_b->value.type())) {
+            // cast(t, a) - cast(t, b) -> cast(t, a - b)
```

Comment:
I recall that we went back and forth on this, but what was this rule necessary for? It's very problematic. Converting Int64(big_int32) - Int64(big_int32) into a 32-bit subtract followed by an upcast seems aggressively unhelpful for certain types of code. A programmer might have upcast before the subtract specifically because the subtract would overflow.

If this only needs to convert Int32(some_i64) - Int32(other_i64) to Int32(some_i64 - other_i64), then that's always correct, at least, but would be a pessimization for vectors...